### ISSUE TITLE
Molecule v2 design feedback

### ISSUE THREAD

**Issue by alvaroaleman** [metacloud/molecule#714]
Hey @retr0h,

Thanks for all your work on Molecule which I appreciate a lot. It does make Ansible testing easier and has saved me quite a lot of time.

Currently you seem to be working mainly on the v2 branch, which is a good thing because it shows Molecule is moving forward.
However, v2 does also bring some uncertainty because there is not much information available on what changes are coming. The only info I am aware of is this [commment of yours](https://github.com/metacloud/molecule/pull/652#issuecomment-268287931) which looks like v2 will be tailored for a very specific use case of yours.

Therefore I am afraid other use cases will not be supported/covered any longer. I am wondering if you can provide some further details on v2? Is there a process/way to add feature/design suggestions for v2 or a way to support your efforts?

**Comment by retr0h** [comment]
Hi @alvaroaleman.  The Molecule v2 branch is work we have wanted to do for quite some time, but getting those changes into v1 would have proved to be too disruptive.

Molecule v2 is a redesign, but it still holds true to many of the v1 principles.  You are still able to execute the same subcommands as v1.  The two main differences with v2 are the way in which instances are managed and scenarios.

1. Instances are managed with ansible, vs having to write code in molecule.  We are basically writing functionality in Molecule to support functionality Ansible already has.

2. We have wanted scenarios for quite some time.  The details of scenarios are still being worked out, but the ability to create, converge, test, destroy various scenarios is powerful, and a hugely requested feature.

At the moment all of this has been prototyped with Docker and Ansible 2.2.  I am just about feature complete with v2, and v2 supports just about all the v1 functionality.  Will then work on an Ansible Vagrant module, and incorporating that into molecule, at which point I'll prob open it up for beta testing.

So basically, I am trying to move molecule forward with a simpler design, make the config consistent, component level configurability (e.g. disabling lint or passing cli args to lint), and adding much needed scenario based testing.

I'm happy to hear suggestions.  I hope that answer your questions.

P.S. Molecule v1 isn't going away anytime soon.  I have been accepting PRs to keep things moving forward.  I have also been backporting features that come into v1 into v2.

**Comment by retr0h** [comment]
I have also been trying to restructure the docs for v2.

http://molecule.readthedocs.io/en/v2/

**Comment by alvaroaleman** [comment]
Hi,

I like the idea of using Ansible for allocating instances, especially since Ansible has a better error handling handling.

What I didn't quite like about the description in your comment is the recursive structure. Instead of having a single `molecule.yml` there will be a lot of them and they must be organized in a specific structure, together with testplaybooks, verifier config etc. This is harder to distribute and to understand.

What in my opinion Molecule is good at is providing an abstraction layer that is easy to understand. Scenarios are definitely a wanted addition, but I think it would be easier to use if they just became part of `molecule.yml`, e.g. something like this:

```ỳaml
scenarios:
  - scenario_a:
      instances:
        - image: el7
          size: m2.medium
      playbooks:
        - tests/playbook_1.yml
        - tests/playbook_2.yml
```

Internaly, this could then be used to generate the recursive file structure you described.
Whats your opinion on that idea?

**Comment by retr0h** [comment]
> What in my opinion Molecule is good at is providing an abstraction layer that is easy to understand. Scenarios are definitely a wanted addition, but I think it would be easier to use if they just became part of molecule.yml, e.g. something like this:

Regardless, we will need separate directories for tests associated with the scenario, so the end user will need to create a directory structure of some sort.  I don't like wedging lots of files under the root of the role.  Having a molecule directory with scenario based directories and tests under it, makes the most sense.  It may not be as clean as test kitchen, but test kitchen doesn't support multi-node anyways.

I'm thinking #684 will help to clean things up, where we have a minimal molecule.yml inside each scenario.  Molecule also has a `molecule init scenario` subcommand to aid with this.

**Comment by alvaroaleman** [comment]
> I don't like wedging lots of files under the root of the role.

I agree with that and that is certainly not what I want.

My point is: What Molecule provides is an abstraction layer that requires me as a user to only define the parts of my configuration that usually change.I do not have too maintain a huge stack of boilerplate code within the role anymore.

For 90% of the uses cases that changing part is:

* A list of scenerios
* For each scenario a list of instances
* For each instance a template and a sizing definition
* For each scenario a list of playbooks
* For each scenario a list of testconfigurations

I don't quite see why a certain directory structure should be enforced. It certainly makes sense to provide sane defaults, but the end user should have the ability to configure something else if he or she wishes to.

This is especially true since one directory per scenario means playbooks and testconfigurations can not be reused.

A very common use case for scenarios will be to have a simple scenario the developer runs on his/her laptop to try to figure out how to achieve a certain goal and do a basic validation. Then have a bigger scenario that contains a full validation and is run by a CI system on an IaaS/Paas platform.
This use case would now require to either duplicate code (CI scenario is the developer scenario + additional scenarios that are run on top of the developer scenario) or to use symlinks.

This does work but it contradicts the initial goal of simplifying the testing configuration as much as possible.

As a sidenote, if we have scenarios Molecule can perfectly well be used for testing of Ansible projects (As in playbooks+roles+vars).

**Comment by retr0h** [comment]
Your concerns have been noted.  I can't guarantee they will be implemented, but have noted your feedback.  Thanks as always!
